storage/objects.go

```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
 	return nil
 }
 
+func abortMpu(c ObjectsClient, ctx context.Context, input *AbortMpuInput) error {
+	reqInput := client.RequestInput{
+		Method:  http.MethodPost,
+		Path:    input.PartsDirectoryPath + "/abort",
+		Headers: &http.Header{},
+		Body:    nil,
+	}
+	respBody, _, err := c.client.ExecuteRequestStorage(ctx, reqInput)
+	if err != nil {
+		return errors.Wrap(err, "unable to abort mpu")
+	}
+
+	if respBody != nil {
+		defer respBody.Close()
+	}
+
+	return nil
+}
+
+func commitMpu(c ObjectsClient, ctx context.Context, input *CommitMpuInput) error {
+	headers := &http.Header{}
+	for key, value := range input.Headers {
+		headers.Set(key, value)
+	}
+
+	id := input.Id
+	idLength := len(id)
+	prefixLen, err := strconv.Atoi(id[idLength-1 : idLength])
+	if err != nil {
+		return errors.Wrap(err, "unable to commit mpu")
+	}
+	prefix := id[:prefixLen]
+	partPath := "/" + c.client.AccountName + "/uploads/" + prefix + "/" + input.Id + "/commit"
+
+	reqInput := client.RequestInput{
+		Method:  http.MethodPost,
+		Path:    partPath,
+		Headers: headers,
+		Body:    input.Body,
+	}
+	respBody, _, err := c.client.ExecuteRequestStorage(ctx, reqInput)
+	if err != nil {
+		return errors.Wrap(err, "unable to commit mpu")
+	}
+
+	if respBody != nil {
+		defer respBody.Close()
+	}
+
+	return nil
+}
+
+func createMpu(c ObjectsClient, ctx context.Context, input *CreateMpuInput) (*CreateMpuOutput, error) {
+	absPath := absFileInput(c.client.AccountName, input.Body.ObjectPath)
+
+	// Because some clients will be treating Manta like S3, they will
+	// include slashes in object names which we'll need to convert to
+	// directories
+	if input.ForceInsert {
+		absDirName := _AbsCleanPath(path.Dir(string(absPath)))
+		exists, _ := checkDirectoryTreeExists(c, ctx, absDirName)
+		if !exists {
+			err := createDirectory(c, ctx, absDirName)
+			if err != nil {
+				return nil, errors.Wrap(err, "unable to create directory for create mpu operation")
+			}
+		}
+	}
+	headers := &http.Header{}
+	for key, value := range input.Body.Headers {
+		headers.Set(key, value)
+	}
+	if input.DurabilityLevel != 0 {
+		headers.Set("Durability-Level", strconv.FormatUint(input.DurabilityLevel, 10))
+	}
+	if input.ContentLength != 0 {
+		headers.Set("Content-Length", strconv.FormatUint(input.ContentLength, 10))
+	}
+	if input.ContentMD5 != "" {
+		headers.Set("Content-MD5", input.ContentMD5)
+	}
+
+	input.Body.ObjectPath = string(absPath)
+	reqInput := client.RequestInput{
+		Method:  http.MethodPost,
+		Path:    "/" + c.client.AccountName + "/uploads",
+		Headers: &http.Header{},
```

Comment:
should this be `Headers: headers` here - looks like your passing in a new empty headers object?

Reply:
Good catch!